src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -321,138 +411,121 @@ int MPL_gavl_tree_free(MPL_gavl_tree_t gavl_tree)
-                        MPL_free(cur_node);
-                        STACK_PUSH(node_stack, inorder_node);
-                    }
-                } else {
-                    inorder_node = cur_node->right;
-                    STACK_PUSH(node_stack, cur_node);
-                    while (inorder_node->left) {
-                        STACK_PUSH(node_stack, inorder_node);
-                        inorder_node = inorder_node->left;
-                    }
-
-                    if (inorder_node->parent != cur_node) {
-                        if (inorder_node->right)
-                            inorder_node->right->parent = inorder_node->parent;
-                        inorder_node->parent->left = inorder_node->right;
-                    } else {
-                        /* only right child of deleted node */
-                        cur_node->right = NULL;
-                    }
-                    val = (void *) cur_node->val;
-                    cur_node->addr = inorder_node->addr;
-                    cur_node->len = inorder_node->len;
-                    cur_node->val = inorder_node->val;
-                    MPL_free(inorder_node);
-                }
-
-                STACK_POP(node_stack, cur_node);
-
-              stack_recovery:
-                gavl_update_node_info(cur_node);
-
-                int lheight = cur_node->left == NULL ? 0 : cur_node->left->height;
-                int rheight = cur_node->right == NULL ? 0 : cur_node->right->height;
-                if (lheight - rheight > 1) {
-                    gavl_tree_node_s *lnode = cur_node->left;
-                    int llheight = lnode->left == NULL ? 0 : lnode->left->height;
-                    if (llheight + 1 == lheight)
-                        gavl_right_rotation(cur_node, lnode);
-                    else
-                        gavl_left_right_rotation(cur_node, lnode);
-                } else if (rheight - lheight > 1) {
-                    gavl_tree_node_s *rnode = cur_node->right;
-                    int rlheight = rnode->left == NULL ? 0 : rnode->left->height;
-                    if (rlheight + 1 == rheight)
-                        gavl_right_left_rotation(cur_node, rnode);
-                    else
-                        gavl_left_rotation(cur_node, rnode);
-                }
-
-                if (!STACK_EMPTY(node_stack)) {
-                    STACK_POP(node_stack, cur_node);
-                    goto stack_recovery;
-                } else {
-                    break;
-                }
-            } while (1);
-
-            while (gavl_tree_iptr->root && gavl_tree_iptr->root->parent != NULL)
-                gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
-
-            if (val && gavl_tree_iptr->gavl_free_fn)
-                gavl_tree_iptr->gavl_free_fn(val);
         }
-    } while (val);
+
+        /* delete the matched node and return corresponding buffer object val */
+        gavl_tree_delete_internal(tree_ptr, dnode, &val);
+
```

Comment:
Why do you return the buffer object but not directly free it inside `gavl_tree_delete_internal `?